### PR TITLE
Avoid tests from failing fast

### PR DESCRIPTION
**Pull request by Rajalakshmi-Girish**
Currently, usage of make test and running the various test suites like unit, integration, e2e are terminating at the frist instance of failure and holding back the possibility of running the entire list of tests.

- Add a gotest flag "-failfast=false " when the mode is not fail_fast - [Link](https://github.com/etcd-io/etcd/pull/15698/files#diff-82ab8ed1742ea8058193b0f3fd5d895c4fe4c21bde8dba50d7389d97536dda9bR318-R321) **DONE** via https://github.com/etcd-io/etcd/pull/15790

- Corrected the logic for the `test.sh` to keep_going with other suites passed to PASSES variable, even on failure with one suite. - [Link](https://github.com/etcd-io/etcd/pull/15698/files#diff-52afcc694c0ff8ff253441f42ff44d6b4734fe30ae9070f2ba07a668a10e7baaL649-L659) **DONE** via https://github.com/etcd-io/etcd/pull/15813

- Not exiting the script when there is failure with one modules out of the list of modules - [Link](https://github.com/etcd-io/etcd/pull/15698/files#diff-82ab8ed1742ea8058193b0f3fd5d895c4fe4c21bde8dba50d7389d97536dda9bR204-R217)

- To make even parallel mode tests to not fail fast - [Link](https://github.com/etcd-io/etcd/pull/15698/files#diff-82ab8ed1742ea8058193b0f3fd5d895c4fe4c21bde8dba50d7389d97536dda9bL322-L323)

- function `integration_pass` and `e2e_pass` had couple of function call and made sure all of them are run before exiting the flow. - [Link](https://github.com/etcd-io/etcd/pull/15698/files#diff-52afcc694c0ff8ff253441f42ff44d6b4734fe30ae9070f2ba07a668a10e7baaL118-L128)

### CODE REVIEW

**Comment by Rajalakshmi-Girish** [comment]
@serathius Can you please take a look?
@mkumatag ^^

File: scripts/test.sh [hunk 1]
```diff
@@ -116,15 +116,21 @@ function integration_extra {
 }
 
 function integration_pass {
-  run_for_module "tests" go_test "./integration/..." "parallel" : -timeout="${TIMEOUT:-15m}" "${COMMON_TEST_FLAGS[@]}" "${RUN_ARG[@]}" -p=2 "$@" || return $?
-  run_for_module "tests" go_test "./common/..." "parallel" : --tags=integration -timeout="${TIMEOUT:-15m}" "${COMMON_TEST_FLAGS[@]}" -p=2 "${RUN_ARG[@]}" "$@" || return $?
+  run_for_module "tests" go_test "./integration/..." "parallel" : \
```
Comment (fuweid) [inline scripts/test.sh:119]: If this command fails but the following commands are successfully, the L124 return 0. It is not expected. 
Please checkout this https://github.com/etcd-io/etcd/pull/15519#discussion_r1144637951

File: scripts/test.sh [hunk 2]
```diff
@@ -100,7 +100,7 @@ function run_unit_tests {
   local pkgs="${1:-./...}"
   shift 1
   # shellcheck disable=SC2086
-  GOLANG_TEST_SHORT=true go_test "${pkgs}" "parallel" : -short -timeout="${TIMEOUT:-3m}" "${COMMON_TEST_FLAGS[@]}" "${RUN_ARG[@]}" "$@"
+  GOLANG_TEST_SHORT=true go_test "${pkgs}" "parallel" : -p=2 -short -timeout="${TIMEOUT:-3m}" "${COMMON_TEST_FLAGS[@]}" "${RUN_ARG[@]}" "$@"
```
Comment (jmhbnz) [inline scripts/test.sh:103]: Thanks for raising these ideas @Rajalakshmi-Girish.

I'm curious on the reasons behind setting parallel to `-p=2` across the board. From memory the default setting is `GOMAXPROCS` if not specified.  So are we saying that this default setting is causing problems and `-p=2` is somehow going to work better? If so I am keen to understand this more.

**Comment by serathius** [comment]
There are a lot of unrelated changes done in this PR. Don't know why we are setting `-p 2` while changing `--failfast` flag balue. Please propose minimal change you want to introduce and move others so separate PRs.

File: scripts/test.sh (hunk 1 above)
Comment (Rajalakshmi-Girish) [reply scripts/test.sh:119]: @fuweid I thought the "return $?" below this command would return the AND of exit codes of all commands.

Thanks for the reference link. I shall correct and re-submit a patch.

**Comment by Rajalakshmi-Girish** [comment]
> There are a lot of unrelated changes done in this PR. Don't know why we are setting `-p 2` while changing `--failfast` flag value. Please propose minimal change you want to introduce and move others so separate PRs.

I have removed the change to add `-p=2` flag. I dint know that tests of packages run in parallel by default using the GOMAXPROCS value.

But the other changes are all related to stop the **failing fast** nature of the test.sh and test_lib.sh scripts.
PTAL at the description https://github.com/etcd-io/etcd/pull/15698#issue-1661803035

**Comment by Rajalakshmi-Girish** [comment]
@serathius Please let me know if you still think I need to break down this PR into multiple smaller PRs.

File: scripts/test.sh (hunk 1 above)
Comment (Rajalakshmi-Girish) [reply scripts/test.sh]: @fuweid Can you please review again?

File: scripts/test.sh (hunk 2 above)
Comment (Rajalakshmi-Girish) [reply scripts/test.sh]: @jmhbnz Can you please review now?

**Comment by Rajalakshmi-Girish** [comment]
@serathius Can you PTAL?

**Comment by jmhbnz** [comment]
Hey @Rajalakshmi-Girish can we close this one now that we have done some individual smaller pull requests?

**Comment by Rajalakshmi-Girish** [comment]
> Hey @Rajalakshmi-Girish can we close this one now that we have done some individual smaller pull requests?

Sure, will close this now.

I was referring to the description https://github.com/etcd-io/etcd/pull/15698#issue-1661803035 and updating the newer PR links to this issue.
Hope this issue's description would be editable even after closing.

**Pull request closed without merging**